File: app/src/main/java/com/trevorwiebe/trackacow/presentation/fragments/MoveFragment.java
package com.trevorwiebe.trackacow.presentation.fragments;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.trevorwiebe.trackacow.R;
import com.trevorwiebe.trackacow.domain.adapters.ShufflePenAndLotsAdapter;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.pen.QueryAllPens;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.lot.QueryLots;
import com.trevorwiebe.trackacow.data.entities.LotEntity;
import com.trevorwiebe.trackacow.data.entities.PenEntity;
import com.trevorwiebe.trackacow.domain.objects.ShuffleObject;
import com.trevorwiebe.trackacow.domain.utils.DragHelper;

import java.util.ArrayList;

public class MoveFragment extends Fragment implements
        QueryAllPens.OnPensLoaded,
        QueryLots.OnLotsLoaded {

    private RecyclerView mMoveRv;
    private TextView mEmptyMoveList;

    private ArrayList<PenEntity> mPenEntities = new ArrayList<>();
    private ShufflePenAndLotsAdapter mShuffleAdapter;
    private Context mContext;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_move, container, false);

        mMoveRv = rootView.findViewById(R.id.shuffle_rv);
        mEmptyMoveList = rootView.findViewById(R.id.empty_move_tv);

        return rootView;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.mContext = context;
    }

    @Override
    public void onResume() {
        super.onResume();
        mMoveRv.setLayoutManager(new LinearLayoutManager(mContext));

        mShuffleAdapter = new ShufflePenAndLotsAdapter();

        DragHelper dragHelper = new DragHelper(mShuffleAdapter);

        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(dragHelper);
        mShuffleAdapter.setTouchHelper(itemTouchHelper);

        mMoveRv.setAdapter(mShuffleAdapter);

        itemTouchHelper.attachToRecyclerView(mMoveRv);

        new QueryAllPens(MoveFragment.this).execute(mContext);
    }

    @Override
    public void onPensLoaded(ArrayList<PenEntity> penEntitiesList) {
        mPenEntities = penEntitiesList;
        new QueryLots(MoveFragment.this).execute(mContext);
    }

    @Override
    public void onLotsLoaded(ArrayList<LotEntity> lotEntities) {
        ArrayList<ShuffleObject> shuffleObjects = new ArrayList<>();
        for (int r = 0; r < mPenEntities.size(); r++) {

            PenEntity penEntity = mPenEntities.get(r);
            String penName = penEntity.getPenName();
            String penId = penEntity.getPenCloudDatabaseId();

            ShuffleObject penShuffleObject = new ShuffleObject(ShufflePenAndLotsAdapter.PEN_NAME, penName, penId);
            shuffleObjects.add(penShuffleObject);

            ArrayList<LotEntity> selectedLotEntities = getLotEntities(penEntity.getPenCloudDatabaseId(), lotEntities);
            if (selectedLotEntities.size() != 0) {
                for (int p = 0; p < selectedLotEntities.size(); p++) {
                    LotEntity lotEntity = selectedLotEntities.get(p);
                    String lotName = lotEntity.getLotName();
                    String lotId = lotEntity.getLotCloudDatabaseId();
                    ShuffleObject lotShuffleObject = new ShuffleObject(ShufflePenAndLotsAdapter.LOT_NAME, lotName, lotId);
                    shuffleObjects.add(lotShuffleObject);
                }
            }

        }

        if (shuffleObjects.size() == 0) {
            mEmptyMoveList.setVisibility(View.VISIBLE);
        } else {
            mEmptyMoveList.setVisibility(View.INVISIBLE);
        }
        mShuffleAdapter.setAdapterVariables(shuffleObjects, mContext);

    }

    private ArrayList<LotEntity> getLotEntities(String penId, ArrayList<LotEntity> lotEntities) {
        ArrayList<LotEntity> selectedLotEntities = new ArrayList<>();
        for (int o = 0; o < lotEntities.size(); o++) {
            LotEntity lotEntity = lotEntities.get(o);
            if (lotEntity.getLotPenCloudDatabaseId().equals(penId)) {
                selectedLotEntities.add(lotEntity);
            }
        }
        return selectedLotEntities;
    }
}

File: app/src/main/java/com/trevorwiebe/trackacow/presentation/fragment_move/MoveFragment.kt
package com.trevorwiebe.trackacow.presentation.fragment_move

import android.content.Context
import androidx.recyclerview.widget.RecyclerView
import android.widget.TextView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.trevorwiebe.trackacow.R
import androidx.recyclerview.widget.LinearLayoutManager
import com.trevorwiebe.trackacow.presentation.fragment_move.utils.DragHelper
import androidx.recyclerview.widget.ItemTouchHelper
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.trevorwiebe.trackacow.data.mapper.compound_mapper.toLotModel
import com.trevorwiebe.trackacow.data.mapper.compound_mapper.toPenModel
import com.trevorwiebe.trackacow.domain.utils.DataSource
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class MoveFragment : Fragment() {

    private lateinit var mMoveRv: RecyclerView
    private lateinit var mEmptyMoveList: TextView
    private lateinit var mProgressBar: LinearProgressIndicator
    private var mObjectList: MutableList<Any> = mutableListOf()
    private lateinit var mShuffleAdapter: ShufflePenAndLotsAdapter

    private val moveViewModel: MoveViewModel by viewModels()

    private lateinit var mContext: Context

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val rootView = inflater.inflate(R.layout.fragment_move, container, false)
        mMoveRv = rootView.findViewById(R.id.shuffle_rv)
        mEmptyMoveList = rootView.findViewById(R.id.empty_move_tv)
        mProgressBar = rootView.findViewById(R.id.move_progress_indicator)
        mMoveRv.layoutManager = LinearLayoutManager(mContext)

        mShuffleAdapter = ShufflePenAndLotsAdapter()

        mShuffleAdapter.onItemShuffled { lotModel ->
            moveViewModel.onEvent(MoveUiEvents.OnItemShuffled(lotModel))
        }

        mShuffleAdapter.onLotsMerged { lotModelList ->
            moveViewModel.onEvent(MoveUiEvents.OnLotsMerged(lotModelList))
        }

        val dragHelper = DragHelper(mShuffleAdapter)
        val itemTouchHelper = ItemTouchHelper(dragHelper)
        mShuffleAdapter.setTouchHelper(itemTouchHelper)

        mMoveRv.adapter = mShuffleAdapter

        itemTouchHelper.attachToRecyclerView(mMoveRv)

        lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                moveViewModel.uiState.collect {

                    mObjectList.clear()

                    if (it.isFetchingFromCloud && it.dataSource == DataSource.Local && it.penAndLotList.isEmpty()) {
                        mProgressBar.visibility = View.VISIBLE
                    } else {
                        mProgressBar.visibility = View.INVISIBLE
                    }

                    val penAndLotList = it.penAndLotList

                    for (r in penAndLotList.indices) {

                        val penAndLotModel = penAndLotList[r]

                        val penModel = penAndLotModel.toPenModel()
                        mObjectList.add(penModel)

                        if (!penAndLotModel.lotName.isNullOrEmpty()) {
                            val lotModel = penAndLotModel.toLotModel()
                            mObjectList.add(lotModel)
                        }
                    }

                    if (mObjectList.size == 0) {
                        Handler(Looper.getMainLooper()).postDelayed(
                            {
                                if (mObjectList.size == 0) {
                                    mEmptyMoveList.visibility = View.VISIBLE
                                }
                            }, 100 // value in milliseconds
                        )
                    } else {
                        mEmptyMoveList.visibility = View.GONE
                    }
                    mShuffleAdapter.setShuffleObjectList(mObjectList, mContext)
                }
            }
        }
        return rootView
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mContext = context
    }
}